client/exec.go

```diff
@@ -0,0 +1,305 @@
+		GroupID     *int              `json:"group-id"`
+		Group       string            `json:"group"`
+		Terminal    bool              `json:"terminal"`
+		Stderr      bool              `json:"stderr"`
+		Width       int               `json:"width"`
+		Height      int               `json:"height"`
+	}{
+		Command:     opts.Command,
+		Environment: opts.Environment,
+		WorkingDir:  opts.WorkingDir,
+		Timeout:     timeoutStr,
+		UserID:      opts.UserID,
+		User:        opts.User,
+		GroupID:     opts.GroupID,
+		Group:       opts.Group,
+		Terminal:    opts.Terminal,
+		Stderr:      opts.Stderr,
+		Width:       opts.Width,
+		Height:      opts.Height,
+	}
+	var body bytes.Buffer
+	err := json.NewEncoder(&body).Encode(&payload)
+	if err != nil {
+		return "", err
+	}
+	headers := map[string]string{
+		"Content-Type": "application/json",
+	}
+	resultBytes, changeID, err := client.doAsyncFull("POST", "/v1/exec", nil, headers, &body)
+	if err != nil {
+		return "", err
+	}
+	var result struct {
+		WebsocketIDs map[string]string `json:"websocket-ids"`
+	}
+	err = json.Unmarshal(resultBytes, &result)
+	if err != nil {
+		return "", err
+	}
+
+	if args == nil {
+		return changeID, nil
+	}
+
+	// Process additional arguments (connecting I/O and websockets)
+	fds := result.WebsocketIDs
+
+	// Call the control handler with a connection to the control socket
+	if args.Control != nil && fds["control"] != "" {
+		conn, err := client.getChangeWebsocket(changeID, fds["control"])
+		if err != nil {
+			return "", err
+		}
+		go args.Control(conn)
+	}
+
+	if opts.Terminal {
+		// Handle terminal-based executions
+		if args.Stdin != nil && args.Stdout != nil {
+			// Connect to the websocket
+			conn, err := client.getChangeWebsocket(changeID, fds["io"])
+			if err != nil {
+				return "", err
+			}
+
+			// And attach stdin and stdout to it
+			go func() {
+				wsutil.WebsocketSendStream(conn, args.Stdin, -1)
+				<-wsutil.WebsocketRecvStream(args.Stdout, conn)
+				conn.Close()
+				if args.DataDone != nil {
+					close(args.DataDone)
+				}
+			}()
+		} else {
+			if args.DataDone != nil {
+				close(args.DataDone)
+			}
+		}
+	} else {
+		// Handle non-terminal executions
+		dones := map[string]chan bool{}
+		conns := []*websocket.Conn{}
+
+		// Handle stdin and stdout
+		if fds["io"] != "" {
+			conn, err := client.getChangeWebsocket(changeID, fds["io"])
+			if err != nil {
+				return "", err
+			}
+			conns = append(conns, conn)
+			dones["stdin"] = wsutil.WebsocketSendStream(conn, args.Stdin, -1)
+			dones["stdout"] = wsutil.WebsocketRecvStream(args.Stdout, conn)
+		}
+
+		// Handle stderr separately if needed
+		if opts.Stderr && fds["stderr"] != "" {
+			conn, err := client.getChangeWebsocket(changeID, fds["stderr"])
+			if err != nil {
+				return "", err
+			}
+			conns = append(conns, conn)
+			dones["stderr"] = wsutil.WebsocketRecvStream(args.Stderr, conn)
+		}
+
+		// Wait for everything to be done
+		go func() {
```

Comment:
Comment seems misplaced since `go func()` won't wait for anything.  I'm also not 100% sure I understand the logic here as my browser is playing games with the tab indentations, but it looks like some code paths allow the command to continue in the background, while others wait for the commands to finish before returning. Unless this is an explicit choice, we shouldn't change the behavior in those terms as a side effect of other options. In other words, the presence of stdin/stdout/stderr should not change whether this function blocks on the command or not. But again, I'm stating that as a conversation starter.. I don't quite follow the design here in those terms, so it may be that I'm just missing something.

Reply:
Good call on all counts. I've updated the comment and simplified the logic so there's only one clear code path.